pkg/validation/formats.go

```diff
@@ -0,0 +1,91 @@
+package validation
+
+import (
+	"errors"
+	jsonschemaformat "github.com/iawaknahc/jsonschema/pkg/jsonschema/format"
+	"github.com/skygeario/skygear-server/pkg/core/phone"
+	"net/mail"
+	"net/url"
+	"path/filepath"
+)
+
+func init() {
+	jsonschemaformat.DefaultChecker["phone"] = FormatPhone{}
+	jsonschemaformat.DefaultChecker["email"] = FormatEmail{AllowName: false}
+	jsonschemaformat.DefaultChecker["email-name-addr"] = FormatEmail{AllowName: true}
+	jsonschemaformat.DefaultChecker["uri"] = FormatURI{}
+}
+
+// FormatPhone checks if input is a phone number in E.164 format.
+// If the input is not a string or is an empty string, it is not an error.
+// To enforce string or non-empty string, use other JSON schema constructs.
+// This design allows this format to validate optional phone number.
+type FormatPhone struct{}
+
+func (f FormatPhone) CheckFormat(value interface{}) error {
+	str, ok := value.(string)
+	if !ok {
+		return nil
+	}
+	return phone.EnsureE164(str)
+}
+
+// FormatEmail checks if input is an email address.
+// If the input is not a string or is an empty string, it is not an error.
+// To enforce string or non-empty string, use other JSON schema constructs.
+// This design allows this format to validate optional email.
+type FormatEmail struct {
+	AllowName bool
+}
+
+func (f FormatEmail) CheckFormat(value interface{}) error {
+	s, ok := value.(string)
+	if !ok {
+		return nil
+	}
+
+	addr, err := mail.ParseAddress(s)
+	if err != nil {
+		return err
+	}
+
+	if !f.AllowName && addr.Name != "" {
+		return errors.New("input email must not have name")
+	}
+
+	return nil
+}
+
+// FormatURI checks if input is an absolute URI, without query/fragment.
```

Comment:
Why `uri` cannot have query or fragment?

Reply:
This is `URLFullOnly` from the original format. I suppose I'll make it accepts any URI (except URI references).